torch/csrc/autograd/variable.h

```diff
@@ -642,12 +603,19 @@ inline const Variable& as_variable_ref(const at::Tensor& tensor) {
   return static_cast<const Variable&>(tensor);
 }
 
-inline const at::Tensor& Variable::data() const noexcept {
-  return get()->data_;
+inline at::Tensor Variable::tensor_data() const noexcept {
+  auto self_impl_copy = get()->shallow_copy_and_detach(
+    /*version_counter=*/get()->version_counter(),
+    /*allow_tensor_metadata_change=*/get()->allow_tensor_metadata_change());
+  return at::Tensor(self_impl_copy);
 }
 
-inline at::Tensor& Variable::data() noexcept {
-  return get()->data_;
+inline at::Tensor Variable::variable_data() const noexcept {
+  auto self_impl_copy = get()->shallow_copy_and_detach(
+    /*version_counter=*/0,
+    /*allow_tensor_metadata_change=*/false);
+  self_impl_copy->set_autograd_meta(c10::guts::make_unique<Variable::AutogradMeta>(self_impl_copy.get(), false));
```

Comment:
what's going on here?  Why is the autograd meta different than in the `tensor_data` case?

Reply:
The legacy `data()` function's semantics is to return a Variable's underlying data tensor (which has no AutogradMeta), and `tensor_data()` tries to mimic this semantics by returning a shallow-copy of the Variable's TensorImpl without setting AutogradMeta.

The `variable_data()` function has the same semantics as the Python `.data` function, which create a new Variable that shares the same storage and tensor metadata with the original Variable, but with a completely new autograd history. (The original `THPVariable_get_data()` function calls `auto var = make_variable(self->cdata.data(), /*requires_grad=*/false, /*allow_tensor_metadata_change=*/false)`, which is equivalent to the code in this function. The new version of `THPVariable_get_data()` just calls this function https://github.com/pytorch/pytorch/pull/17072/files#diff-13b395c368197e2cac32474f78c76ae4R209)